File: codes/Garnetwzy/85.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        if(matrix.size() == 0)
            return 0;
        int n = matrix[0].size();
        vector<int> height(n, 0);
        vector<int> left(n, 0);
        vector<int> right(n, n);
        int maxVal = 0;
        for(int i = 0; i < matrix.size(); i++) {
            int curLeft = 0, curRight = n;
            for(int j = 0; j < n; j++) {
                if(matrix[i][j] == '1') {
                    height[j]++;
                    left[j] = max(left[j], curLeft);
                }else{
                    height[j] = 0;
                    left[j] = 0;
                    curLeft = j+1;
                }
            }
            for(int j = n-1; j >= 0; j--) {
                if(matrix[i][j] == '1') {
                    right[j] = min(right[j], curRight);
                }else{
                    right[j] = n;
                    curRight = j;
                }
            }
            for(int i = 0; i < n; i++) {
                maxVal = max(maxVal, height[i] * (right[i] - left[i]));
            }
        }
        return maxVal;
    }
};

File: codes/HellDream/85.java
class Solution {
    public int maximalRectangle(char[][] matrix) {
        if(matrix.length==0||matrix[0].length==0) return 0;
        int[] heights = new int[matrix[0].length];
        int max = 0;
        for(int i=0;i<matrix.length;i++){
            for(int j=0;j<matrix[i].length;j++){
                char c = matrix[i][j];
                if(c=='0'){
                    heights[j] = 0;
                }else{
                    heights[j] = heights[j]+1;
                }
            }
            max = Math.max(max, largestRectangleArea(heights));
        }
        return max;
    }
    
    public int largestRectangleArea(int[] heights) {
        if(heights.length==0) return 0;
        int max = 0;
        for(int i=0;i<heights.length;i++){
            if(i == heights.length-1|| heights[i]>heights[i+1]){
                int minHeight = heights[i];
                for(int j=i;j>=0;j--){
                    minHeight = Math.min(minHeight,heights[j]);
                    int width = i-j+1;
                    max = Math.max(max, minHeight*width);
                }
            }
        }
        return max;
    }
}